core/domain/suggestion_services.py

```diff
@@ -0,0 +1,382 @@
+    """Assigns a user to review the suggestion.
+
+    Args:
+        suggestion: Suggestion. The domain object of the suggestion to be
+            updated.
+        assigned_reviewer_id: str. The ID of the user who is assigned to review.
+    """
+    suggestion.status = suggestion_models.STATUS_IN_REVIEW
+    suggestion.assigned_reviewer_id = assigned_reviewer_id
+
+    _update_suggestion(suggestion)
+
+
+# TODO (nithesh): This function is temporary. At the moment, the feedback
+# threads id is of the form exp_id.<random_str> while the suggestion ids are of
+# the form entity_type.entity_id.<random_str>. Once the feedback thread ID
+# migration is complete, these two IDs will be matched, and this function will
+# be removed.
+def get_thread_id_from_suggestion_id(suggestion_id):
+    """Gets the thread_id from the suggestion_id.
+
+    Args:
+        suggestion_id: str. The ID of the suggestion.
+
+    Returns:
+        str. The thread ID linked to the suggestion.
+    """
+    return suggestion_id[suggestion_id.find('.') + 1:]
+
+
+def is_suggestion_handled(suggestion):
+    """Checks if the suggestion has been accepted or rejected.
+
+    Args:
+        suggestion: Suggestion. The domain object of the suggestion to be
+            checked.
+
+    Returns:
+        bool. Whether the suggestion has been handled or not.
+    """
+    return (suggestion.status not in [suggestion_models.STATUS_IN_REVIEW,
+                                      suggestion_models.STATUS_RECEIVED])
+
+
+def get_commit_message_for_suggestion(author_username, commit_message):
+    """Returns a modified commit message for an accepted suggestion.
+
+    Args:
+        author_username: str. Username of the suggestion author.
+        commit_message: str. The original commit message submitted by the
+            suggestion author.
+
+    Returns:
+        str. The modified commit message to be used in the exploration commit
+        logs.
+    """
+    return '%s %s: %s' % (
+        feconf.COMMIT_MESSAGE_ACCEPTED_SUGGESTION_PREFIX,
+        author_username, commit_message)
+
+
+def accept_suggestion(suggestion, reviewer_id, commit_message):
+    """Accepts the given suggestion after validating it.
+
+    Args:
+        suggestion: Suggestion. The domain object of the suggestion to be
+            accepted.
+        reviewer_id: str. The ID of the reviewer accepting the suggestion.
+        commit_message: str. The commit message.
+
+    Raises:
+        Exception: The suggestion is already handled.
+        Exception: The suggestion is not valid.
+        Exception: The commit message is empty.
+    """
+    if is_suggestion_handled(suggestion):
+        raise Exception('The suggestion has already been accepted/rejected.')
+    if not commit_message or not commit_message.strip():
+        raise Exception('Commit message cannot be empty.')
+
+    try:
+        # If an exception is raised, the suggestion status will be set as
```

Comment:
That's a bit weird. I think it needs to be more granular than this, and perhaps a separate message defined for this check. What exactly is the concern here?

If it's e.g. a state name changing then have a separate method devoted to actually checking that and other stuff that you want to check at this juncture. Otherwise you end up shadowing errors that should really be caught and fixed.

Reply:
With reference to the [above](https://github.com/oppia/oppia/pull/4982#discussion_r192581023) comment, I think a general validation of the suggestion needs to be done before accepting to make sure that everything is fine. And yes I think this process does not actually tell you what the error is and instead just says the suggestion is valid. 

For the moment I don't have a design for getting suggestions out of an invalid status. But I wanted to mark status as STATUS_INVALID and this status change should be brought about if an exception is raised in the validate(). This was the best approach I had in mind, any suggestions for a better approach.

Reply:
We could have a cron job that iterates through suggestions and invalidates them periodically, but that feels like overkill. I think the question really is _when_ to mark a suggestion invalid and I think it needs to be initiated by someone -- so, if not due to a commit, it must be initiated by whoever is triaging the suggestion.

So I think for now it's reasonable to let the triager actually reject the suggestion manually along the lines of my comment above. Validity checks prior to regular saves of the suggestion object should be reserved for actual structural integrity (like the suggestion is formatted correctly and makes sense as a suggestion object) but they do not need to check referential integrity. The latter is checked on the fly in the frontend and at the time of acceptance of a suggestion.

Reply:
Ok I see your point. So basically while updating suggestions we need to validate that parameter X is actually a string and so on. But for the second point, don't we actually need a validation in the backend too? To check that the suggestion references valid properties/values to be edited? Yeah this probably should go into a separate validator like a pre_accept_validator or similar. 

If we want to invalidate/validate when commits are made. An important question arises -- If a state is added with name same as a previously deleted state, should all suggestions to the previously deleted state become valid again? I would go with your suggested approach of greying out the accept button with a message saying "the suggestion can't be accepted because of XYZ reason please ask the author to make a new suggestion to a valid state."

Reply:
Yes, we do need backend validation for referential integrity but this only really matters on suggestion acceptance. So we can do it then.

I like the message you suggest, except that you might not need the "please ask the author" part. You could prepopulate the rejection message to explain what happened though. Remember that the suggester cannot see the state and doesn't even know the state name.

Good question about a state being added with the same name as a previously deleted state. It's hard to handle this, short of using state IDs (cf. @prasanna08's work), but it doesn't really matter too much since the creator is looking at the old/new content anyway. So I think it's fine if the invalid comments become valid again and subsequently get rejected.

Reply:
Well, actually, "fine" is a strong word. It's suboptimal, but I don't see a way around it except to listen to commits, which seems expensive.

Reply:
So shall we get rid of STATUS_INVALID concept altogether? and have the reject message indicate that it was invalid?

Reply:
I don't know what you _should_ do (probably want to ask @anmolshkl who has more context), but I have no immediate objection to what you said.

Reply:
I agree with @seanlip. It is better to leave the validation of structural integrity till the end i.e. when a suggestion is accepted/rejected. Also, I don't think the renaming of states could be a major problem since the decision would be made by the reviewer as to whether the suggestion is valid or not (assuming we will be pulling the latest contents of the state based on its name). If the state name is not found, you could indicate that it is an invalid suggestion or else just show the state contents to the reviewer, in which case, he can decide if the old content and the new content make sense or not.

Reply:
Yeah. Makes sense. I'll separate this check out from validate and also in the frontend, we'll validate the suggestion when displayed to a reviewer and grey out the accept button if not valid. This removes the necessity of the invalid flag to be stored (We can determine it during execution). So I'll remove that too.

Reply:
Removed the invalid status, split the validation into separate functions for structural and referential validity. Also we'll have a check in the frontend as mentioned.